### PR TITLE
Avoid an overflow in proximity detection

### PR DESCRIPTION
**Pull request by cor3ntin**
See #705

### CODE REVIEW

File: targets/playground/examples/thymio-default-behaviours.aesl [hunk 1]
```diff
@@ -343,14 +343,27 @@ if run_behavior == START then
...
-		
+
 	elseif state == EXPLORER then
 		# Breitenberg obtacle avoidance
 		if speed > 0 then
-			temp = (prox.horizontal[0]*1+prox.horizontal[1]*2+prox.horizontal[2]*3+prox.horizontal[3]*2+prox.horizontal[4]*1)
-			temp2 = prox.horizontal[0]*-4+prox.horizontal[1]*-3+prox.horizontal[3]*3+prox.horizontal[4]*4
-			motor.left.target = speed - (temp+temp2) / (2000/speed)
-			motor.right.target = speed - (temp-temp2) / (2000/speed)
+			temp = prox.horizontal[0] * 1
```
Comment (stephanemagnenat) [inline targets/playground/examples/thymio-default-behaviours.aesl]: The easiest would be to use `math.dot`.

File: targets/playground/examples/thymio-default-behaviours.aesl [hunk 2]
```diff
@@ -343,14 +343,27 @@ if run_behavior == START then
...
+				+ prox.horizontal[2] * 3
+				+ prox.horizontal[3] * 2
+				+ prox.horizontal[4] * 1
+
+			temp2 = prox.horizontal[0] * -4
+				+ prox.horizontal[1] * -3
+				+ prox.horizontal[3] * 3
+				+ prox.horizontal[4] * 4
+
+			temp  = temp * 2000/speed
```
Comment (stephanemagnenat) [inline targets/playground/examples/thymio-default-behaviours.aesl]: This might overflow, please us `math.muldiv`.

**Comment by cor3ntin** [comment]
@stephanemagnenat ping

File: targets/playground/examples/thymio-default-behaviours.aesl [hunk 3]
```diff
@@ -341,15 +344,20 @@ if run_behavior == START then
...
 		if speed > 0 then
-			temp = (prox.horizontal[0]*1+prox.horizontal[1]*2+prox.horizontal[2]*3+prox.horizontal[3]*2+prox.horizontal[4]*1)
-			temp2 = prox.horizontal[0]*-4+prox.horizontal[1]*-3+prox.horizontal[3]*3+prox.horizontal[4]*4
-			motor.left.target = speed - (temp+temp2) / (2000/speed)
-			motor.right.target = speed - (temp-temp2) / (2000/speed)
+			call math.dot(temp, kernel1[0:4], prox.horizontal[0:4], 1)
+			call math.dot(temp2, kernel2[0:4], prox.horizontal[0:4], 1)
 			callsub motorbound
+
+			call math.muldiv(temp, 2000, speed)
```
Comment (stephanemagnenat) [inline targets/playground/examples/thymio-default-behaviours.aesl]: `math.muldiv` takes 4 parameters, see https://www.thymio.org/en:asebastdnative. I guess the problem does not compile.

File: targets/playground/examples/thymio-default-behaviours.aesl [hunk 4]
```diff
@@ -81,6 +81,9 @@ var dir
 var rc5_speed_l
 var rc5_speed_t
 
+var kernel1[5] = [1,2,3,2,1]
+var kernel2[5] = [-4, -3, 1, 3, 4]
```
Comment (stephanemagnenat) [inline targets/playground/examples/thymio-default-behaviours.aesl]: Be consistent with spacing between `kernel1` and `kernel2`. Also, maybe there are more relevant names, such as `avoidanceKernelX`, or even by replacing `1` and `2` with their semantic meaning (@mbonani might be able to give some explanation about the semantics).

**Comment by cor3ntin** [comment]
@davidjsherman I haven't access to the ci yet ( I sent you a mail ), so I can't see what is failling

**Comment by stephanemagnenat** [comment]
I'm fixing the bugs and merging by hand.

File: targets/playground/examples/thymio-default-behaviours.aesl [hunk 5]
```diff
@@ -341,15 +344,20 @@ if run_behavior == START then
...
-			motor.left.target = speed - (temp+temp2) / (2000/speed)
-			motor.right.target = speed - (temp-temp2) / (2000/speed)
+			call math.dot(temp,  avoidanceKernel1[0:4], prox.horizontal[0:4], 1)
+			call math.dot(temp2, avoidanceKernel2[0:4], prox.horizontal[0:4], 1)
 			callsub motorbound
+
+			call math.muldiv(temp, temp, 2000, speed)
+			call math.muldiv(temp2, temp2, 2000, speed)
+
+			motor.left.target  = (speed - temp) - temp2
```
Comment (stephanemagnenat) [inline targets/playground/examples/thymio-default-behaviours.aesl]: It is more natural wrt robotics and symmetric with the next line to write `speed - (temp + temp2)`.

File: targets/playground/examples/thymio-default-behaviours.aesl [hunk 6]
```diff
@@ -341,15 +344,20 @@ if run_behavior == START then
...
-		
+
 	elseif state == EXPLORER then
 		# Breitenberg obtacle avoidance
 		if speed > 0 then
-			temp = (prox.horizontal[0]*1+prox.horizontal[1]*2+prox.horizontal[2]*3+prox.horizontal[3]*2+prox.horizontal[4]*1)
-			temp2 = prox.horizontal[0]*-4+prox.horizontal[1]*-3+prox.horizontal[3]*3+prox.horizontal[4]*4
-			motor.left.target = speed - (temp+temp2) / (2000/speed)
-			motor.right.target = speed - (temp-temp2) / (2000/speed)
+			call math.dot(temp,  avoidanceKernel1[0:4], prox.horizontal[0:4], 1)
```
Comment (stephanemagnenat) [inline targets/playground/examples/thymio-default-behaviours.aesl]: The last parameter N should by 0, as we divide by `2^N`.

**Comment by stephanemagnenat** [comment]
Merged by cf2ab7b525a3adc51be3591d0ae144c77e271555 and 6bd019f9d734c524c181630d9ca6f389f321db71.

@mbonani could you test the latest version to see if it works as in the real robot?

**Comment by davidjsherman** [comment]
`math.dot` is cleaner and also protected against overflow, isn't it?

**Comment by stephanemagnenat** [comment]
They do something different, `math.dot` divides by 2^N, while `math.muldiv` divides by N, with N being the last parameters. In both cases the multiplication is carried out in 32 bits, and the result is truncated to 16 bits after the shift/division.

File: targets/playground/examples/thymio-default-behaviours.aesl (hunk 5 above)
Comment (cor3ntin) [reply targets/playground/examples/thymio-default-behaviours.aesl]: Good call.

File: targets/playground/examples/thymio-default-behaviours.aesl (hunk 6 above)
Comment (cor3ntin) [reply targets/playground/examples/thymio-default-behaviours.aesl]: Indeed. A case of not reading the docs carefully enough

**Pull request closed without merging**